plugins/catalog-backend/src/ingestion/processors/microsoftGraph/read.ts

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { GroupEntity, UserEntity } from '@backstage/catalog-model';
+import { buildOrgHierarchy } from '../util/org';
+import { MicrosoftGraphClient } from './client';
+import {
+  MICROSOFT_GRAPH_GROUP_ID_ANNOTATION,
+  MICROSOFT_GRAPH_TENANT_ID_ANNOTATION,
+  MICROSOFT_GRAPH_USER_ID_ANNOTATION,
+} from './constants';
+
+export function normalizeEntityName(name: string): string {
+  return name
+    .trim()
+    .toLocaleLowerCase()
+    .replace(/[^a-zA-Z0-9_\-\.]/g, '_');
+}
+
+export async function readMicrosoftGraphUsers(
+  client: MicrosoftGraphClient,
+  options?: { userFilter?: string },
+): Promise<{
+  users: UserEntity[]; // With all relations empty
+}> {
+  const entities: UserEntity[] = [];
+  const picturePromises: Promise<void>[] = [];
+
+  for await (const user of client.getUsers({
+    filter: options?.userFilter,
+    select: ['id', 'displayName', 'mail', ''],
+  })) {
+    if (!user.id || !user.displayName || !user.mail) {
+      continue;
+    }
+
+    const name = normalizeEntityName(user.mail);
+    const entity: UserEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'User',
+      metadata: {
+        name,
+        annotations: {
+          [MICROSOFT_GRAPH_USER_ID_ANNOTATION]: user.id!,
+        },
+      },
+      spec: {
+        profile: {
+          displayName: user.displayName || undefined,
+          email: user.mail || undefined,
+
+          // TODO: Additional fields?
+          // jobTitle: user.jobTitle || undefined,
+          // officeLocation: user.officeLocation || undefined,
+          // mobilePhone: user.mobilePhone || undefined,
+        },
+        memberOf: [],
+      },
+    };
+
+    // Download the photos in parallel, otherwise it can take quite some time
+    const loadPhoto = async () => {
+      entity.spec.profile!.picture = await client.getUserPhotoWithSizeLimit(
+        user.id!,
+        // We are limiting the photo size, as users with full resolution photos
+        // can make the Backstage API slow
+        120,
```

Comment:
I can imagine. Does not the api expose pictures as https urls, that we can use an auth token to get? Perhaps through a proxy and a service account?

Reply:
See above.

But about the proxy: That would be an option too, only problem is that the proxy-backend isn't flexible enough for the flow that retrieves the access token. We could write a custom proxy backend for that and also move this code their.